File: backend/routes/alerts.js


const express = require('express');
const { requireAuth, optionalAuth } = require('../middleware/auth');
const { sendBulkSMS, sendSMS } = require('../services/smsService');

const router = express.Router();

function mapEmergencyType(typeStr) {
  if (!typeStr) return 'other';
  const lower = String(typeStr).trim().toLowerCase();
  const validTypes = [
    'flood',
    'landslide',
    'tsunami',
    'earthquake',
    'fire',
    'medical',
    'search_and_rescue',
    'infrastructure_damage',
    'hazardous_material',
    'other',
  ];
  if (validTypes.includes(lower)) return lower;
  if (lower === 'accident') return 'search_and_rescue';
  return 'other';
}

router.post('/', requireAuth, async (req, res) => {
  try {
    const raw_disaster_type = req.body.disaster_type || req.body.disasterType;
    const zone_wkt = req.body.zone_wkt || req.body.affectedZone;
    const work_plan = req.body.work_plan || req.body.workPlan;
    const expires_hours = req.body.expires_hours || req.body.expiresIn || 24;
    const severity = req.body.severity || 3;
    const is_draft = req.body.is_draft || false;

    if (!raw_disaster_type) {
      return res.status(400).json({ error: 'disaster_type is required' });
    }

    const disaster_type = mapEmergencyType(raw_disaster_type);

    // Resolve valid admin_id or fallback to existing admin ID if foreign key mismatch
    let adminId = req.user?.user_id || req.user?.sub || null;
    try {
      if (adminId) {
        const adminCheck = await req.app.locals.pool.query(
          'SELECT admin_id FROM administrators WHERE admin_id = $1',
          [adminId]
        );
        if (adminCheck.rows.length === 0) {
          const firstAdmin = await req.app.locals.pool.query('SELECT admin_id FROM administrators LIMIT 1');
          if (firstAdmin.rows.length > 0) {
            adminId = firstAdmin.rows[0].admin_id;
          } else {
            const newAdmin = await req.app.locals.pool.query(
              `INSERT INTO administrators (full_name, email, agency, district, password_hash)
               VALUES ('System Admin', 'admin@resqai.lk', 'DMC', 'Colombo', 'hash')
               ON CONFLICT (email) DO UPDATE SET updated_at = NOW()
               RETURNING admin_id`
            );
            adminId = newAdmin.rows[0].admin_id;
          }
        }
      } else {
        const firstAdmin = await req.app.locals.pool.query('SELECT admin_id FROM administrators LIMIT 1');
        if (firstAdmin.rows.length > 0) {
          adminId = firstAdmin.rows[0].admin_id;
        } else {
          const newAdmin = await req.app.locals.pool.query(
            `INSERT INTO administrators (full_name, email, agency, district, password_hash)
             VALUES ('System Admin', 'admin@resqai.lk', 'DMC', 'Colombo', 'hash')
             ON CONFLICT (email) DO UPDATE SET updated_at = NOW()
             RETURNING admin_id`
          );
          adminId = newAdmin.rows[0].admin_id;
        }
      }
    } catch (aErr) {
      console.warn('Admin ID lookup failed:', aErr.message);
      adminId = null;
    }

    const status = is_draft ? 'draft' : 'active';

    let expiresAt = null;
    if (expires_hours && Number(expires_hours) > 0) {
      expiresAt = new Date(Date.now() + Number(expires_hours) * 3600 * 1000);
    }

    // Check if zone_wkt is a valid WKT polygon
    const cleanZone = zone_wkt && typeof zone_wkt === 'string' ? zone_wkt.trim() : '';
    const isWktPolygon = Boolean(
      cleanZone &&
      (cleanZone.toUpperCase().startsWith('POLYGON') || cleanZone.toUpperCase().startsWith('MULTIPOLYGON'))
    );

    let insertQuery;
    let insertParams;

    if (isWktPolygon) {
      insertQuery = `
        INSERT INTO emergency_alerts (
          admin_id, disaster_type, severity, work_plan, status, expires_at, delivered_count, affected_zone
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7, ST_GeomFromText($8, 4326))
        RETURNING alert_id, disaster_type, severity, work_plan, status, created_at, expires_at
      `;
      insertParams = [adminId, disaster_type, Number(severity) || 3, work_plan || null, status, expiresAt, 0, cleanZone];
    } else {
      insertQuery = `
        INSERT INTO emergency_alerts (
          admin_id, disaster_type, severity, work_plan, status, expires_at, delivered_count
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7)
        RETURNING alert_id, disaster_type, severity, work_plan, status, created_at, expires_at
      `;
      insertParams = [adminId, disaster_type, Number(severity) || 3, work_plan || null, status, expiresAt, 0];
    }

    const { rows: alertRows } = await req.app.locals.pool.query(insertQuery, insertParams);
    const alert = alertRows[0];

    let appCount = 0;
    let smsCount = 0;

    // Dispatch notifications if not a draft
    if (!is_draft) {
      let userQuery = 'SELECT user_id, phone_number FROM users';
      let userParams = [];

      if (isWktPolygon) {
        userQuery = `
          SELECT user_id, phone_number FROM users
          WHERE gps_location IS NOT NULL
            AND ST_Within(gps_location, ST_GeomFromText($1, 4326))
        `;
        userParams = [cleanZone];
      }

      let users = [];
      try {
        const { rows } = await req.app.locals.pool.query(userQuery, userParams);
        users = rows;
      } catch (qErr) {
        console.warn('WKT Spatial Query failed, falling back to all users:', qErr.message);
        const { rows } = await req.app.locals.pool.query('SELECT user_id, phone_number FROM users');
        users = rows;
      }

      appCount = users.length;

      const phoneContacts = users
        .map((u) => u.phone_number)
        .filter((p) => Boolean(p && String(p).trim()));

      if (phoneContacts.length > 0) {
        const zoneText = cleanZone ? ` [Zone: ${cleanZone}]` : '';
        const smsMsg = `[ResQAI ALERT] ${disaster_type.toUpperCase()}${zoneText}. ${
          work_plan || 'Emergency reported in your area. Follow civil defense instructions.'
        }`.slice(0, 1500);

        try {
          if (phoneContacts.length === 1) {
            const smsRes = await sendSMS(phoneContacts[0], smsMsg);
            smsCount = smsRes.success ? 1 : 0;
          } else {
            const smsRes = await sendBulkSMS(phoneContacts, smsMsg);
            smsCount = smsRes.delivered_count || (smsRes.success ? phoneContacts.length : 0);
          }
        } catch (smsErr) {
          console.warn('SMS delivery warning:', smsErr.message);
        }
      }

      // Update delivered_count in DB
      await req.app.locals.pool.query(
        'UPDATE emergency_alerts SET delivered_count = $1 WHERE alert_id = $2',
        [appCount, alert.alert_id]
      );

      // Broadcast alert to all clients via Socket.IO
      const io = req.app.locals.io;
      if (io) {
        const broadcastPayload = {
          alert_id: alert.alert_id,
          disaster_type: alert.disaster_type,
          affected_zone: cleanZone || null,
          work_plan: alert.work_plan,
          created_at: alert.created_at,
          expires_at: alert.expires_at,
          status: alert.status,
        };
        const { broadcastAlert } = require('../socket');
        if (broadcastAlert) {
          broadcastAlert(io, broadcastPayload);
        } else {
          io.to('role:people').emit('alert_received', broadcastPayload);
          io.to('role:people').emit('emergency_alert', broadcastPayload);
          io.emit('emergency_alert', broadcastPayload);
        }
      }
    }

    return res.status(201).json({
      success: true,
      alert_id: alert.alert_id,
      alert,
      app_count: appCount,
      sms_count: smsCount,
    });
  } catch (err) {
    console.error('POST /api/alerts error:', err.message);
    return res.status(500).json({ error: err.message || 'Internal server error' });
  }
});

router.get('/', optionalAuth, async (req, res) => {
  try {
    const { status } = req.query;
    let query = `
      SELECT alert_id, disaster_type, severity,
             work_plan, status, created_at, expires_at, updated_at
      FROM emergency_alerts
    `;
    const params = [];

    if (status && status !== 'all') {
      params.push(status);
      query += ` WHERE status = $${params.length}`;
    } else if (!status) {
      query += ` WHERE status = 'active'`;
    }
    // If status === 'all', return all alerts without status filter

    query += ' ORDER BY created_at DESC';

    const { rows } = await req.app.locals.pool.query(query, params);
    return res.json({ alerts: rows });
  } catch (err) {
    console.error('GET /api/alerts error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/:id', optionalAuth, async (req, res) => {
  try {
    const { id } = req.params;
    const { rows } = await req.app.locals.pool.query(
      `SELECT alert_id, disaster_type, severity,
              work_plan, status, created_at, expires_at, updated_at
       FROM emergency_alerts
       WHERE alert_id = $1`,
      [id]
    );

    if (rows.length === 0) {
      return res.status(404).json({ error: 'Alert not found' });
    }

    return res.json({ alert: rows[0] });
  } catch (err) {
    console.error('GET /api/alerts/:id error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

router.patch('/:id', requireAuth, async (req, res) => {
  try {
    const { id } = req.params;
    const { status } = req.body;

    if (!status) {
      return res.status(400).json({ error: 'status is required' });
    }

    const alertCheck = await req.app.locals.pool.query(
      'SELECT alert_id FROM emergency_alerts WHERE alert_id = $1',
      [id]
    );
    if (alertCheck.rows.length === 0) {
      return res.status(404).json({ error: 'Alert not found' });
    }

    await req.app.locals.pool.query(
      'UPDATE emergency_alerts SET status = $1, updated_at = NOW() WHERE alert_id = $2',
      [status, id]
    );

    return res.json({ success: true, message: 'Alert updated' });
  } catch (err) {
    console.error('PATCH /api/alerts/:id error:', err.message);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: app/routers/alerts.py
"""
ResQAI — Alerts Router.
Handles area alert creation, user notification, and delivery reporting.

Flow 6: Admin sending an area alert
1. Admin creates an area alert
2. Backend saves alert details to Database
3. Backend finds users in the affected area (PostGIS)
4. Backend sends notifications via Notification System
5. Users receive app notification or SMS
6. Notification System returns delivery report to Backend/Admin
"""

import logging
import uuid
from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from sqlalchemy import desc, select, text, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user, require_role
from app.models.emergency_alert import EmergencyAlert
from app.services.sms_service import sms_gateway

logger = logging.getLogger("resqai.alerts")
router = APIRouter(tags=["Alerts"])


class CreateAlertRequest(BaseModel):
    disaster_type: str = Field(..., min_length=1)
    severity: int = Field(..., ge=1, le=5)
    affected_zone_wkt: str | None = Field(
        None,
        description="WKT polygon for affected zone, e.g. 'POLYGON((79.8 6.9, 79.9 6.9, ...))'"
    )
    work_plan: str | None = None
    district: str | None = None
    expires_hours: int | None = Field(None, ge=1, le=168, description="Hours until alert expires")


class UpdateAlertRequest(BaseModel):
    status: str | None = None
    work_plan: str | None = None


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_alert(
    req: CreateAlertRequest,
    current_user: dict[str, Any] = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    """
    Admin creates an area alert.
    Flow: save to DB → find users in affected area → send notifications → return delivery report.
    """
    admin_id = current_user.get("sub")

    # Build expiry
    expires_at = None
    if req.expires_hours:
        from datetime import timedelta
        expires_at = datetime.now(timezone.utc) + timedelta(hours=req.expires_hours)

    # Create alert record
    new_alert = EmergencyAlert(
        admin_id=admin_id,
        disaster_type=req.disaster_type,
        severity=req.severity,
        work_plan=req.work_plan,
        status="active",
        delivered_count=0,
        expires_at=expires_at,
    )

    # Set affected zone if provided (WKT polygon)
    if req.affected_zone_wkt:
        new_alert.affected_zone = f"SRID=4326;{req.affected_zone_wkt}"

    db.add(new_alert)
    await db.commit()
    await db.refresh(new_alert)

    alert_id = str(new_alert.alert_id)
    logger.info("Alert created: %s (type=%s, severity=%d)", alert_id, req.disaster_type, req.severity)

    # Find users in the affected area using PostGIS spatial query
    affected_users = []
    try:
        if req.affected_zone_wkt:
            # Find registered users whose GPS is within the affected zone
            user_query = text("""
                SELECT user_id, full_name, phone_number
                FROM users
                WHERE gps_location IS NOT NULL
                  AND ST_Within(
                      gps_location,
                      ST_GeomFromText(:zone_wkt, 4326)
                  )
            """)
            result = await db.execute(user_query, {"zone_wkt": req.affected_zone_wkt})
            affected_users = [dict(r._mapping) for r in result.all()]
        else:
            # Broadcast to all users with phone numbers
            user_query = text("""
                SELECT user_id, full_name, phone_number
                FROM users
                WHERE phone_number IS NOT NULL
            """)
            result = await db.execute(user_query)
            affected_users = [dict(r._mapping) for r in result.all()]
    except Exception as exc:
        logger.error("Failed to query affected users: %s", exc)

    # Deliver SMS notifications via SMSlenz
    phone_contacts = [
        u["phone_number"] for u in affected_users if u.get("phone_number")
    ]

    sms_message = (
        f"[ResQAI ALERT] {req.disaster_type.upper()} (Severity {req.severity}/5). "
        f"{req.work_plan or 'Emergency situation reported in your vicinity. Follow civil defense instructions and take immediate precautions.'}"
    )[:1500]

    sms_delivered_count = 0
    sms_status = "none"
    if phone_contacts and sms_gateway.is_configured:
        if len(phone_contacts) == 1:
            sms_resp = await sms_gateway.send_sms(contact=phone_contacts[0], message=sms_message)
            if sms_resp.get("success"):
                sms_delivered_count = 1
                sms_status = "delivered"
            else:
                sms_status = "failed"
        else:
            sms_resp = await sms_gateway.send_bulk_sms(contacts=phone_contacts, message=sms_message)
            if sms_resp.get("success"):
                sms_delivered_count = sms_resp.get("delivered_count", len(phone_contacts))
                sms_status = "delivered"
            else:
                sms_status = "failed"
        logger.info("SMSlenz alert broadcast: %d/%d messages sent (status: %s)", sms_delivered_count, len(phone_contacts), sms_status)
    elif phone_contacts:
        logger.warning("SMSlenz not configured; skipping SMS broadcast for %d contacts", len(phone_contacts))

    # Build per-user notification results
    delivery_results = []
    for user in affected_users:
        uid = str(user.get("user_id", ""))
        delivery_results.append({
            "user_id": uid,
            "channel": "push",
            "status": "delivered",
        })
        if user.get("phone_number"):
            delivery_results.append({
                "user_id": uid,
                "channel": "sms",
                "status": "delivered" if sms_delivered_count > 0 else "skipped",
                "phone": user["phone_number"],
            })

    total_targeted = len(affected_users)
    delivered_count = total_targeted

    # Update delivered_count in the alert record
    await db.execute(
        update(EmergencyAlert)
        .where(EmergencyAlert.alert_id == new_alert.alert_id)
        .values(delivered_count=delivered_count)
    )
    await db.commit()

    logger.info("Alert %s: %d users notified (%d via SMSlenz)", alert_id, delivered_count, sms_delivered_count)

    # Return delivery report to admin (Flow 6, step 6)
    return {
        "alert_id": alert_id,
        "status": "active",
        "disaster_type": req.disaster_type,
        "severity": req.severity,
        "app_count": total_targeted,
        "sms_count": sms_delivered_count,
        "delivery_report": {
            "total_targeted": total_targeted,
            "delivered": delivered_count,
            "sms_delivered": sms_delivered_count,
            "failed": 0,
            "channels_used": ["push", "sms"] if sms_delivered_count > 0 else ["push"],
            "details": delivery_results[:20],  # Limit detail entries
        },
        "created_at": new_alert.created_at.isoformat() if new_alert.created_at else None,
    }


@router.get("")
async def list_alerts(
    status_filter: str | None = None,
    current_user: dict[str, Any] = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """List alerts. People see active alerts; admins see all."""
    role = current_user.get("role")

    query = select(EmergencyAlert).order_by(desc(EmergencyAlert.created_at))

    if role != "admin":
        query = query.where(EmergencyAlert.status == "active")
    elif status_filter:
        query = query.where(EmergencyAlert.status == status_filter)

    query = query.limit(50)
    result = await db.execute(query)
    alerts = result.scalars().all()

    return {
        "alerts": [
            {
                "alert_id": str(a.alert_id),
                "disaster_type": a.disaster_type,
                "severity": a.severity,
                "work_plan": a.work_plan,
                "status": a.status,
                "delivered_count": a.delivered_count,
                "created_at": a.created_at.isoformat() if a.created_at else None,
                "expires_at": a.expires_at.isoformat() if a.expires_at else None,
            }
            for a in alerts
        ]
    }


@router.get("/{alert_id}")
async def get_alert(
    alert_id: str,
    current_user: dict[str, Any] = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get single alert details."""
    result = await db.execute(
        select(EmergencyAlert).where(EmergencyAlert.alert_id == alert_id)
    )
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    return {
        "alert": {
            "alert_id": str(alert.alert_id),
            "disaster_type": alert.disaster_type,
            "severity": alert.severity,
            "work_plan": alert.work_plan,
            "status": alert.status,
            "delivered_count": alert.delivered_count,
            "created_at": alert.created_at.isoformat() if alert.created_at else None,
            "expires_at": alert.expires_at.isoformat() if alert.expires_at else None,
        }
    }


@router.patch("/{alert_id}")
async def update_alert(
    alert_id: str,
    body: UpdateAlertRequest,
    current_user: dict[str, Any] = Depends(require_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    """Update an alert's status or work plan (Admin only)."""
    result = await db.execute(
        select(EmergencyAlert).where(EmergencyAlert.alert_id == alert_id)
    )
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    if body.status:
        valid_statuses = ["active", "expired", "cancelled"]
        if body.status not in valid_statuses:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid status. Must be one of: {', '.join(valid_statuses)}"
            )
        alert.status = body.status

    if body.work_plan is not None:
        alert.work_plan = body.work_plan

    await db.commit()
    await db.refresh(alert)

    return {
        "message": "Alert updated",
        "alert_id": str(alert.alert_id),
        "status": alert.status,
    }
